### ISSUE TITLE
Specify email credentials via command line parameters

### ISSUE THREAD

**Issue by kriswilk** [rss2email/rss2email#219]
I'd like to store/manage my `rss2email.cfg` file in GitHub. However, I don't want to commit my email username/password.

Issue #5 relates to feeds with auth credentials and suggests a separate file for these. That may in fact be desirable for feed-related credentials, since a user might have quite a few. I look forward to work on that issue, even though it doesn't currently affect me.

For outgoing email credentials, I propose more urgently adding command line parameters `--username` and `--password` which would override any value set in the config file for the active sending method.

I could then store my credentials in environment variables (as I do for many other applications), and my crontab can just include something like:

```
*/10 * * * * r2e run --username $R2E_USERNAME --password $R2E_PASSWORD
```

Thoughts? Is there a way to accomplish this already that I've just missed?

**Comment by auouymous** [comment]
That would be easier than implementing and migrating to an auth file. It is however a security issue because other users on the system can see the environment and arguments of all commands by default.

**Comment by amiryal** [comment]
> I'd like to store/manage my `rss2email.cfg` file in GitHub.

In case there are other ways to achieve the same goal, would you mind sharing the full user story?

As an example, for sharing your list of feeds with the world, `r2e opmlexport` is the more traditional way to go.

**Comment by auouymous** [comment]
Try running `ps a`, `cat /proc/$$/cmdline |strings`, and `cat /proc/$$/environ |strings`. The /proc directory contains a sub-directory for each process, which has `cmdline` and `environ` files. Any user can watch for changes to /proc and read files inside the sub-directories.

You can mount proc with hidepid=2 to prevent non-wheel users from seeing processes for other users. I don't know if /proc/&lt;pid&gt;/environ is 400 for all systems, a new "feature", or a result of something I've done, but /proc/&lt;pid&gt;/cmdline would allow anyone to see your second crontab example password.

#200 removed the broken interpolation in the config file, but maybe we could add our own for some values and treat a new auth file (~/.config/rss2email.auth) like the process' environment. The interpolation would only happen at the use site to prevent storing the interpolated value back into the config which is what happened before. The auth file can be read at startup and a function wraps any use site that wants to use the auth file. And a double $$ can escape a dollar sign to avoid another #197.

```
smtp-username = ${SMTP_USERNAME}
[feed.1]
url = http://${FEED1_USERNAME}:${FEED1_PASSWORD}@domain.com
```

This would eliminate any migration issues when adding the auth file because it would be opt-in, instead of something that tries to extract existing credentials from the config and move them to the auth file.

**Comment by kriswilk** [comment]
> > I'd like to store/manage my `rss2email.cfg` file in GitHub.
> 
> In case there are other ways to achieve the same goal, would you mind sharing the full user story?
> 
> As an example, for sharing your list of feeds with the world, `r2e opmlexport` is the more traditional way to go.

I keep the configurations for all the applications that run on my server in GitHub to make recovering or moving to a new machine as painless as possible. It also helps me to track changes I've made to configurations easily.

All my other applications optionally allow credentials to be passed in via environment variables and/or command-line options. So it seems like a prudent enhancement for rss2email.

**Comment by kriswilk** [comment]
> Try running `ps a`, `cat /proc/$$/cmdline |strings`, and `cat /proc/$$/environ |strings`. The /proc directory contains a sub-directory for each process, which has `cmdline` and `environ` files. Any user can watch for changes to /proc and read files inside the sub-directories.

Thanks for this... a security expert I am not! I guess my first instinct was right: to load and pass in environment variables.

> ```
> smtp-username = ${SMTP_USERNAME}
> [feed.1]
> url = http://${FEED1_USERNAME}:${FEED1_PASSWORD}@domain.com
> ```
> 
> This would eliminate any migration issues when adding the auth file because it would be opt-in, instead of something that tries to extract existing credentials from the config and move them to the auth file.

This solution sounds great and would have a much broader scope than what I had in mind. I'm all for it. But I suspect it will take considerably more work. Perhaps there's a place for both? I suspect command-line options (used properly) would satisfy most users, and might be a good first step to abstracting private data from the configuration.

If I have some time today or tomorrow I'll look at the code and take a stab at it.

**Comment by kriswilk** [comment]
Continuing to think out loud...

Since command-line options could inadvertently leak sensitive information, how about simply checking for two environment variables such as `RSS2EMAIL_USERNAME` and `RSS2EMAIL_PASSWORD` and letting them take precedence if they are set?

**Comment by amiryal** [comment]
Based on reading [systemd.exec(5)](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#LoadCredential=ID:PATH), I get the impression that it was pretty well thought out, and the solution that follows from it is in this function:
https://github.com/rss2email/rss2email/blob/c9516d345770266c21f95d6eea4339a0894326fd/rss2email/feeds.py#L197
Looks like it only needs to add `_os.environ.get('CREDENTIALS_DIRECTORY')` in the most significant (last) position, if set, and that’s it!

The way to use it is, create a separate directory (mode `700`) and `rss2email.cfg` in it populated with secret configuration, set the environment variable `CREDENTIALS_DIRECTORY` to the path of that directory when invoking `r2e`, and configparser will merge the values from it into the final configuration.

The way to use it with systemd is similar, but instead of setting the environment variable directly, it should be configured with `LoadCredential=rss2email.cfg:/path/to/credentials_directory`.

**Comment by amiryal** [comment]
> Wouldn't merging a "private" config file like this cause commands that write to the configuration to dump the private information into the primary config file?

Good point, well asked! If anything, the main configuration would be written to the private file. However, it would not happen in this case, because `CREDENTIALS_DIRECTORY` should only be set for `r2e run` and not for any command that writes the configuration back. To be extra safe, remove write permissions from the file before invoking.

**Comment by kriswilk** [comment]
> Good point, well asked! If anything, the main configuration would be written to the private file. However, it would not happen in this case, because `CREDENTIALS_DIRECTORY` should only be set for `r2e run` and not for any command that writes the configuration back. To be extra safe, remove write permissions from the file before invoking.

Fair enough. But then there's another kink: commands like `list`, `add` and `delete` will be broken if they cannot touch the secret configuration.

Although it's more work, I think @auouymous's suggestion to reintroduce some form of interpolation is preferable in the long-term.

In the short-term, I have just proposed a much simpler solution that focuses on my original issue. See #220.

**Comment by auouymous** [comment]
I still think the best approach is to have feeds, imap and smtp call a function to interpolate a value and not write it back to config. The interpolator can pull user-defined key values from a fixed location auth file or from an auth file defined in an environment variable. It works for anything now and in the future requiring confidential data, unlike #220 which only supports smpt/imap and would become redundant if this is every implemented. It also clearly indicates settings in the config file that use credentials unlike using configparser to load a second config file.

**Comment by adigitoleo** [comment]
This would be a great benefit. I currently can't get SMTP to work at all, because my SMTP username would contain an `@`, which seems to cause trouble (domain not found, I think it tries to use the second half of the username as the SMTP server domain). I tried replacing it with `%40` as I do in my aerc config, but still no luck (authentication failure). The only thing I can think of in the second case is that my password contains an `#` which is causing the rest of the line in the .cfg file to become ignored.
